File: polytope_feature/utility/list_tools.py
def bisect_left_cmp(arr, val, cmp):
    left = -1
    r = len(arr)
    while r - left > 1:
        e = (left + r) >> 1
        if cmp(arr[e], val):
            left = e
        else:
            r = e
    return left


def bisect_right_cmp(arr, val, cmp):
    left = -1
    r = len(arr)
    while r - left > 1:
        e = (left + r) >> 1
        if cmp(arr[e], val):
            left = e
        else:
            r = e
    return r

File: rust/src/list_tools.rs
pub fn bisect_left_cmp<T, F>(arr: &[T], val: &T, mut cmp: F) -> isize
where
    F: FnMut(&T, &T) -> bool,
{
    let mut left: isize = -1;
    let mut right: isize = arr.len() as isize;

    while right - left > 1 {
        let e = (left + right) >> 1;
        let elem = &arr[e as usize];

        if cmp(elem, val) {
            left = e;
        } else {
            right = e;
        }
    }

    left
}
